machine-learning/ml/pipelines/binary.py

```diff
@@ -39,48 +65,82 @@ def run(self):
         for dataset in self._datasets:
             log("Dataset {}".format(dataset))
 
+            # test set github, validation set github
             for refactoring in self._refactorings:
                 refactoring_name = refactoring.name()
                 log("**** Refactoring Type: %s" % refactoring_name)
 
-                features, x, y, scaler = retrieve_labelled_instances(dataset, refactoring)
-                # test if any refactorings were found for the given refactoring type
-                if x is None:
-                    print("Skip model building for refactoring type: " + refactoring.name())
-                    log("Skip model building for refactoring type: " + refactoring.name())
-                    continue
-
-                # we split in train and test
-                # (note that we use the same split for all the models)
-                x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=TEST_SPLIT_SIZE, random_state=42)
-
-                for model in self._models_to_run:
-                    model_name = model.name()
-
-                    try:
-                        log("Model {}".format(model.name()))
-                        self._start_time()
-                        test_scores, model_to_save = self._run_single_model(model, x, y, x_train, x_test, y_train, y_test)
-
-                        # log test scores
-                        log(format_test_results(dataset, refactoring_name, model_name + " test", test_scores["precision"],
-                                           test_scores["recall"], test_scores['accuracy'], test_scores['tn'],
-                                           test_scores['fp'], test_scores['fn'], test_scores['tp']))
-
-                        # we save the best estimator we had during the search
-                        model.persist(dataset, refactoring_name, features, model_to_save, scaler)
-
-                        self._finish_time(dataset, model, refactoring)
-                    except Exception as e:
-                        print(e)
-                        print(str(traceback.format_exc()))
-
-                        log("An error occurred while working on refactoring " + refactoring_name + " model " + model.name())
-                        log(str(e))
-                        log(str(traceback.format_exc()))
-
-
-    def _run_single_model(self, model_def, x, y, x_train, x_test, y_train, y_test):
+                # we have two options to select a test set,
+                # 1.) Predefined in the database
+                if TEST_SPLIT_SIZE < 0 and len(VALIDATION_DATASETS) > 0:
+                    train_features, x_train, y_train, scaler = retrieve_labelled_instances(dataset, refactoring, True)
```

Comment:
should we extract the content of this `if` and then the `else`, and put in two separated methods? 

`train_test_split_per_project` and `train_test_split_per_percentage` or something like that?

Reply:
I don't see the benefit, there is little duplicate code and the pipeline would not get any easier.